### PR TITLE
[gardener/gardener#2462] Add nodeSelector to run only on allowed worker pools

### PR DESCRIPTION
**Pull request by jannickfahlbusch**
**How to categorize this PR?**
/priority normal
/area networking

**What this PR does / why we need it**:
This makes the the following components run only on pools that are allowed to host system components and extensions:
* calico-kube-controller
* calico-node-vertical-autoscaler
* calico-typha-deploy
* calico-typha-horizontal-autoscaler
* calico-typha-vertical-autoscaler

**Which issue(s) this PR fixes**:
Relates to https://github.com/gardener/gardener/issues/2462
Needs https://github.com/gardener/gardener/pull/2480

**Special notes for your reviewer**:

**Release note**:

```improvement operator
All non-DaemonSet resources now respect the node label `worker.gardener.cloud/system-components=true` as introduced by https://github.com/gardener/gardener/pull/2480
```

### CODE REVIEW

**Comment by jannickfahlbusch** [comment]
/hold

This can only be merged after https://github.com/gardener/gardener/pull/2480 got merged

**Comment by zanetworker** [comment]
> Hm, if we merge it like this then the Calico extension would not be compatible with older Gardener versions anymore. Probably it's safer to let the reconciler check the respective `Shoot`, and only add the node selector in case it detects that it has worker pool that explicitly allows system components (this way it knows that the shoot was reconciled by a new Gardener version).

Was about to say that this won't work if the workers do not have that label.

**Comment by jannickfahlbusch** [comment]
This would need to be handled here right?

https://github.com/gardener/gardener-extension-networking-calico/blob/d7c117e2b59d31dcd5ac57684eab718e9d52d53b/pkg/charts/utils.go#L114-L139

**Comment by zanetworker** [comment]
But since you would need to fetch the value from the shoot, you need access to the cluster object: 
https://github.com/gardener/gardener-extension-networking-calico/blob/d7c117e2b59d31dcd5ac57684eab718e9d52d53b/pkg/controller/actuator_reconcile.go#L53 

Then pass it to compute the chart values.

**Comment by zanetworker** [comment]
/lgtm

@rfranzke since you reviewed this initially, do you want to have another look?

**Comment by jannickfahlbusch** [comment]
We need to wait for the next release of gardener/gardener to vendor the changes here right?

**Comment by vpnachev** [comment]
You can vendor even the master, but a new minor release is planned to be released soon, so waiting for it is also an option.

**Comment by zanetworker** [comment]
@jannickfahlbusch Okay I see you are making use of `SystemComponentsAllowed`, so yes but functionality wise there is no dependencies.

File: charts/internal/calico/templates/typha/deployment-calico-typha.yaml [hunk 1]
```diff
@@ -49,6 +49,9 @@ spec:
     spec:
       nodeSelector:
         beta.kubernetes.io/os: linux
```
Comment (vpnachev) [inline charts/internal/calico/templates/typha/deployment-calico-typha.yaml:51]: I see you have removed this selector for the `calico-kube-controller`, is it possible to remove it also for `calico-typha`?

File: charts/internal/calico/templates/kube-controller/deployment-calico-kube-controllers.yaml [hunk 2]
```diff
@@ -34,7 +34,7 @@ spec:
         gardener.cloud/role: system-component
     spec:
       nodeSelector:
-        beta.kubernetes.io/os: linux
+{{ toYaml .Values.nodeSelector | indent 8 }}
```
Comment (vpnachev) [inline charts/internal/calico/templates/kube-controller/deployment-calico-kube-controllers.yaml:37]:
```suggestion
      {{- if .Values.nodeSelector }}
      nodeSelector:
{{ toYaml .Values.nodeSelector | indent 8 }}
      {{- end }}
```

You can omit the node selector if it is not set in the values.

File: charts/internal/calico/values.yaml [hunk 3]
```diff
@@ -24,4 +24,6 @@ images:
   calico-kube-controllers: "image-repository:image-tag"
   calico-podtodaemon-flex: "image-repository:image-tag"
   calico-cpva: "image-repository:image-tag"
-  calico-cpa: "image-repository:image-tag"
\ No newline at end of file
+  calico-cpa: "image-repository:image-tag"
+
+nodeSelector: {}
```
Comment (vpnachev) [inline charts/internal/calico/values.yaml:29]: If you decide to implement the above suggestion, you have to comment this line.
```suggestion
# nodeSelector: {}
```

File: charts/internal/calico/templates/typha/deployment-calico-typha.yaml (hunk 1 above)
Comment (jannickfahlbusch) [reply charts/internal/calico/templates/typha/deployment-calico-typha.yaml:51]: I've removed it by accident. I'll readd it

File: charts/internal/calico/values.yaml (hunk 3 above)
Comment (jannickfahlbusch) [reply charts/internal/calico/values.yaml]: Done

File: charts/internal/calico/templates/kube-controller/deployment-calico-kube-controllers.yaml (hunk 2 above)
Comment (jannickfahlbusch) [reply charts/internal/calico/templates/kube-controller/deployment-calico-kube-controllers.yaml:38]: Done

File: charts/internal/calico/templates/typha/deployment-calico-typha.yaml (hunk 1 above)
Comment (jannickfahlbusch) [reply charts/internal/calico/templates/typha/deployment-calico-typha.yaml]: Removed the selector

**Comment by jannickfahlbusch** [comment]
@vpnachev @zanetworker I've worked in the changes.

Please take a look again

File: pkg/charts/utils.go [hunk 4]
```diff
@@ -135,6 +136,13 @@ func ComputeCalicoChartValues(network *extensionsv1alpha1.Network, config *calic
 		},
 		"config": calicoConfig,
 	}
+
+	if workerSystemComponentsActivated {
+		calicoChartValues["nodeSelector"] = map[string]string{
+			gardenv1beta1constants.LabelWorkerPoolSystemComponents: "True",
```
Comment (rfranzke) [inline pkg/charts/utils.go:142]: This should be `true`

File: pkg/charts/charts_test.go [hunk 5]
```diff
@@ -353,9 +353,57 @@ var _ = Describe("Chart package test", func() {
...
+						},
+					},
+					"ipv4": map[string]interface{}{
+						"pool":                string(poolIPIP),
+						"mode":                string(always),
+						"autoDetectionMethod": nil,
+					},
+				},
+				"nodeSelector": map[string]string{
+					"worker.gardener.cloud/system-components": "True",
```
Comment (rfranzke) [inline pkg/charts/charts_test.go:398]: `true`

**Pull request merged**